### DISCUSSION TITLE
🆕 Loki Query Builder Feedback

### DISCUSSION THREAD

**Discussion by ivanahuckova** [grafana/grafana#50785] [Observability]
In Grafana v9 we have added a new **Loki visual query builder** as a new beta feature :tada: :sparkles:  ! 

The goal of this query builder is to allow anyone to create, edit and understand LogQL queries even if you are a new Loki user who is not yet familiar with the syntax and mechanics of LogQL. At least that is the goal. It might take us some time to reach this goal, and your feedback could really help in shaping its future.

### What it can do

* Select label filters from simple dropdowns.
* Add functions, aggregations and log operations via the operations selector.
* Supports binary operations on queries.
* Parser that enabled going between the Code (text) mode and builder mode without losing changes.
* Explain mode that breaks the query down and tries to explain what each part does. 
* A new auto legend format mode. 
  * This is very basic in this first iteration. If your results only contain a single label key then the series names will contain the label values.
 
<img width="1418" alt="image" src="https://user-images.githubusercontent.com/30407135/173586990-cd0ac6c3-33c8-4028-8be0-e3e007e74f7c.png">

### Limitations

The builder is not designed to handle all possible LogQL queries. It is designed to handle the most common forms of queries. There should be a warning when going from Code (text) mode to builder mode if the current query cannot be accurately represented in the builder mode. 

### Explain mode 

The explain mode breaks the query down into all its individual parts and functions and brings in documentation and explanations for each part. 

### Query patterns 

This feature is still in a very early state. The thinking is to allow easy access and exploration of common best practice query patterns. Query patterns are currently designed to only replace the operations (functions) of the query but leave your label filters intact. If you have suggestions for this feature and suggestions for good query patterns to include let us know. 

### Feedback

Please let us know what you think. Your feedback could help us decide what to add or change in a future iteration. :sparkles:

**Comment by darwinz** [comment]
How do we do regular, typed-out queries?

**Reply by darwinz** [reply]
Nevermind, I figured it out

**Reply by lindhe** [reply]
I'm also confused about this. How did you find it?

EDIT: Oh, there's a button to the very right of the query, where one can select "Code".

**Comment by malsatin** [comment]
Hi. `Label filter expression` don't have `=~` and `!~` operators in the dropdown
![image](https://user-images.githubusercontent.com/9138285/176990706-5024e0de-f78f-4768-ad67-3e8ee89d0e6e.png)

**Reply by ivanahuckova** [reply]
You are right! Thank you for bringing this up. I have created issue https://github.com/grafana/grafana/issues/51784 and will look into fixing this.

**Reply by ivanahuckova** [reply]
@malsatin the fix was merged and will be part of the next patch release.

**Comment by ydelwiche** [comment]
It may allow more complex searches, but it now takes a lot more clicks to make a simple filter with a few labels.

**Reply by aumkarpraja** [reply]
Agreed. Feels less intuitive going through Loki to look for specific containers and namespaces, a few clicks turned into more clicks.

**Comment by blyefd** [comment]
It looks like the query builder does not support `by ...` when using a Range function operation.
eg:

in the **Code** editor I type the following working query:
`last_over_time({mylabel="myvalue"} | json | unwrap myjsonobj [$__interval]) by (anotherlabel) `

I click on **Builder** and change something (eg a label) and then `by (anotherlabel)`  disappears (in the code preview).

Then I click on **Code** again and confirm it's deleted and the query results are different.

(v9.0.7)

**Reply by ivanahuckova** [reply]
Thanks for bringing this up. I've created and issue out of this https://github.com/grafana/grafana/issues/53570

**Comment by fernferret** [comment]
I'd like to see the ability to have my Metric/Log Query **combined** *in the explore view*.

If I put a metric query in, often I'm looking for a problem, then I'll want to see the logs for that query, like this:

![image](https://user-images.githubusercontent.com/72811/185741051-6b710f8b-5e2a-4bc4-a63e-d969bd344f9d.png)

This already works really well (just by putting 2 queries in) for things like filtering a pod/namespace out, however adding text filters is really difficult as I have to add the same filters, graphically, to both queries, when in reality they really are the same thing, just one is performing metrics on the previous Log Query.

I'm familiar with LogQL but am trying out the new query builder to see how it can help folks that don't know LogQL. I think showing the LogQL query is fantastic under the query and will help folks who are not familiar with LogQL become more proficient.

**Reply by ivanahuckova** [reply]
@fernferret thank you for your feedback. This actually is something we are currently working on. +cc @diegoadams

**Comment by yosiasz** [comment]
Very nice! Would be nice to add a bit more config details for Windows environment for example

```
__path__: D:/lokilogs/*
```

Also is it possible to add multiple folders in path and can they be remote network paths?
```
//sqlserver/logs
//fileserver/logs
//xserver/logs
```

**Comment by aheitzmann** [comment]
I started with a simple query like this on some request logs for one of our services:
```rate({app="proxy", namespace="qa2"} | json  status | status >= 200 | __error__ != `LabelFilterErr` [$__interval])```
When I switch to the query builder, the format of the raw query is correctly maintained. However, any edits through the query builder (e.g. adding a Sum operator) change the status label filter to use a string value:
```sum(rate({app="proxy", namespace="qa2"} | json status | status >= `200` | __error__ != `LabelFilterErr` [$__interval]))```
which results in an error "parse error at line 1, col 67: syntax error: unexpected STRING, expecting BYTES or NUMBER or DURATION"

**Reply by ivanahuckova** [reply]
Hello @aheitzmann, this is actually an issue with loki and the fix was merged https://github.com/grafana/loki/issues/6713 and it should be part of the next release.

**Comment by Rohlik** [comment]
This tool is absolutely impossible to use in my case (Grafana on Linux with latest Chrome) 👎🏽. I'm getting some error:
`parse error at line 1, col 11: syntax error: unexpected IDENTIFIER, expecting = or =~ or !~ or !=`, but this can be just an error on my side.   However I have issue with typing while I switch from `Builder` back to `Code`.  
As you can see, I wanted to replace `logfmt` with `json`, so I select `logfmt` and I start writing and this happened:

https://user-images.githubusercontent.com/2051154/193256911-1c0f7720-4189-4fb2-aa67-e6f67360cabc.mp4

Cursor illogically move to the front from each of my key press 🙊.

**Reply by svennergr** [reply]
Hi @Rohlik,
thanks for your feedback. That is a known issue which is further described here: https://github.com/grafana/grafana/pull/54566#issuecomment-1261230387

It should be fixed in the newest Grafana 9, 8 and 7.

**Comment by Rohlik** [comment]
I found another issue in latest (9.1.6) Grafana. Builder tend to toss values and delete **group by** part from LogQL when you switch from Code to Builder 😞. Check out the video bellow 📹:

https://user-images.githubusercontent.com/2051154/193568208-edd804ca-7d5b-4b29-8a08-b805c20b216a.mp4

**Reply by ivanahuckova** [reply]
Yes, that is a bug already reported in https://github.com/grafana/grafana/issues/53570 and there is already a fix PR https://github.com/grafana/grafana/pull/56184

**Reply by Rohlik** [reply]
Diky. Will it solve also the tossing issue? 🤔

**Reply by ivanahuckova** [reply]
Yes, it will. 🙂 The fix was merged and should be part of next 9.2 release.

**Comment by ArthurGuerquin** [comment]
Hello ! The Loki query generator works well and is easy to use, but I find myself making the same queries over and over and having to type or paste them. Could it be possible to generate a list of query in configuration files and maybe edit a list of custom query templates? 
Thank you very much 😄

**Reply by malsatin** [reply]
> Hello ! The Loki query generator works well and is easy to use, but I find myself making the same queries over and over and having to type or paste them. Could it be possible to generate a list of query in configuration files and maybe edit a list of custom query templates? Thank you very much 😄

Grafana has 'Query History' and 'Starred Query' features, that will probably fit in your case.

**Comment by liguozhong** [comment]
![image](https://user-images.githubusercontent.com/9583245/201870959-1f8f48fb-ef85-4657-93d0-77ee8c449335.png)

add one button (sort by time) when search by uuid or traceid from micro service ，we need analyse log sorted by time. 
thanks . @ivanahuckova

**Reply by malsatin** [reply]
On the right side you can see "Flip results order", it does exactly what you want.

**Reply by liguozhong** [reply]
It was my mistake, grafana loki explore was correct.
The current code is indeed group by (timestamp), I thought it was group by (label, timestamp). 

My loki is out of order because the time of promtail is not the time in the log, but the system time. I will solve this out-of-order problem by introducing a timestamp stage to promtail.

**Comment by gelicia** [comment]
Hello! We want to let you all know that we are moving to Github Issues for feature requests. You can read more about the move [here](https://github.com/grafana/grafana/discussions/73424).

Please feel free to submit issues in github for any further bug reports or feature requests.
